src/middlewared/middlewared/plugins/network.py

```diff
@@ -213,97 +215,85 @@ async def do_update(self, data):
-
-        if not (
-                not await self.middleware.call('system.is_freenas') and
-                await self.middleware.call('failover.licensed')
-        ):
-            for key in ['hostname_virtual', 'hostname_b']:
-                data.pop(key, None)
-            is_ha = False
 
-        new_config.update(data)
         verrors = await self.validate_general_settings(data, 'global_configuration_update')
-        if is_ha and config['hostname_virtual'] != new_config['hostname_virtual']:
-            ad_enabled = await self.middleware.call('activedirectory.get_state') != "DISABLED"
-            if ad_enabled:
+
+        # we need to check if the `hostname_virtual` parameter changed in a couple places
+        # in this method, so go ahead and set it here
+        virt_hostname_changed = False
+        if new_config.get('hostname_b', False):
+            virt_hostname_changed = config['hostname_virtual'] != new_config['hostname_virtual']
+
+        # if this is an HA system and the virtual_hostname has changed we need to make sure
+        # that if the system is joined to AD they leave the domain first and then change
+        # that parameter
+        if virt_hostname_changed:
+            if await self.middleware.call('activedirectory.get_state') != "DISABLED":
                 verrors.add('global_confiugration_update.hostname_virtual',
                             'This parameter may not be changed after joining Active Directory (AD). '
                             'If it must be changed, the proper procedure is to leave the AD domain '
                             'and then alter the parameter before re-joining the domain.')
 
         verrors.check()
 
-        new_config['domains'] = ' '.join(new_config.get('domains', []))
-        new_config['netwait_ip'] = ' '.join(new_config.get('netwait_ip', []))
-        new_config.pop('hostname_local', None)
+        # pop the `hostname_local` key since that's created in the _extend method
+        # and doesn't exist in the database
+        new_hostname = new_config.pop('hostname_local', None)
 
+        # update the db
         await self.middleware.call(
             'datastore.update',
             'network.globalconfiguration',
             config['id'],
             new_config,
             {'prefix': 'gc_'}
         )
-        service_announcement = new_config.pop('service_announcement')
-        new_config['domains'] = new_config['domains'].split()
-        new_config['netwait_ip'] = new_config['netwait_ip'].split()
-
-        def set_(data):
-            data = dict(
-                data,
-                domains=json.dumps(sorted(list(set('domains'))), sort_keys=True),
-                service_announcement=None,
-                netwait_ip=json.dumps(sorted(list(set('netwait_ip'))), sort_keys=True),
-                activity=json.dumps(data['activity'], sort_keys=True),
-            )
-            return set(data.items())
-
-        if len(set_(new_config) ^ set_(config)) > 0:
```

Comment:
Not sure why we had to remove this?

Reply:
A few reasons.

1. prone to breakage if new items are added to API
(for example, service_announcement key was being overwritten to None)
2. potentially expensive
(calling json.dump on all keys in the dict twice for each key that couldn't be flattened normally)
3. the new activity feature was tied to when we restarted hostname/resolvconf which isn't needed. So if you only changed the outbound activity, we would restart hostname/resolvconf?